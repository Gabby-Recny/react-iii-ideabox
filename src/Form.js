import React, { useState } from 'react';
import './Form.css';

const Form = (props) => {
  const [ title, setTitle ] = useState('');
  const [ description, setDescription ] = useState('');

  
    const submitIdea = event => {
      event.preventDefault();
      const newIdea = {
        id: Date.now(),
        title,
        description,
      }
      props.addIdea(newIdea);
      clearInputs();
    }
    
   const clearInputs = () => {
      setTitle('');
      setDescription('');
    }
  return (
    <form>
      <input
        type='text'
        placeholder='Title'
        name='title'
        value={ title }
        onChange={event => setTitle(event.target.value)}
      />

      <input
        type='text'
        placeholder='Description'
        name='description'
        value={ description }
        onChange={event => setDescription(event.target.value)}
      />

    <button onClick={submitIdea}>SUBMIT</button>
  </form>

  )
}


// class Form extends Component {
//   constructor() {
//     super();
//     this.state = {
//       title: '',
//       description: ''
//     }
//   }

//   handleChange = event => {
//     this.setState({[event.target.name]: event.target.value})
//   }


//   render() {
//     return (
//       <form>
//         <input
//           type='text'
//           placeholder='Title'
//           name='title'
//           value={this.state.title}
//           onChange={event => this.handleChange(event)}
//         />

//         <input
//           type='text'
//           placeholder='Description'
//           name='description'
//           value={this.state.description}
//           onChange={event => this.handleChange(event)}
//         />

//         <button onClick={event => this.submitIdea(event)}>SUBMIT</button>
//       </form>
//     )
//   }
// }

export default Form;
