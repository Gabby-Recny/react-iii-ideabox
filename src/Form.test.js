import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';

describe('Form', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Form addIdea={jest.fn()} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when handleChange is called', () => {
    const mockEvent = { target: { name: 'title', value: 'Sweaters for pugs'} };
    const expected = 'Sweaters for pugs';

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state('title')).toEqual(expected);
  });

  it('should reset state when clearInputs is called', () => {
    const expected = {title: '', description: ''};

    wrapper.instance().setState({title: 'Sweaters for pugs', description: 'Why not?'});

    wrapper.instance().clearInputs();

    expect(wrapper.state()).toEqual(expected);
  });

  it('should call addIdea with correct props when submitIdea is called', () => {
    // mock Date.now()
    global.Date.now = jest.spyOn(global.Date, 'now').mockImplementation(() => 12345);

    // set up state with an idea
    const currentState = {title:'Sweaters for pugs', description: 'Because cute'};
    const mockIdea = {...currentState, id: 12345};

    wrapper.instance().setState(currentState);

    // execute submitIdea
    wrapper.instance().submitIdea({preventDefault: jest.fn()});

    // expect addIdea to be called with the mockIdea
    expect(wrapper.instance().props.addIdea).toHaveBeenCalledWith(mockIdea);
  });

  it('should call clearInputs when submitIdea is called', () => {
    wrapper.instance().clearInputs = jest.fn();
    wrapper.instance().submitIdea({preventDefault: jest.fn()});

    expect(wrapper.instance().clearInputs).toHaveBeenCalled();
  });

  it('should run handleChange when the inputs detect a change', () => {
    const mockTitleEvent = { target: { name: 'title', value: 'Sweaters for pugs'} };
    const mockDescEvent = { target: { name: 'description', value: 'because of cuteness'} };

    wrapper.instance().handleChange = jest.fn();
    wrapper.find('[name="title"]').simulate('change', mockTitleEvent);
    wrapper.find('[name="description"]').simulate('change', mockDescEvent);

    expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockTitleEvent);
    expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockDescEvent);
  });

  it('should run submitIdea when the button is clicked', () => {
    const mockEvent = { preventDefault: jest.fn() };

    wrapper.instance().submitIdea = jest.fn();
    wrapper.find('button').simulate('click', mockEvent);

    expect(wrapper.instance().submitIdea).toHaveBeenCalledWith(mockEvent);
  });
});
