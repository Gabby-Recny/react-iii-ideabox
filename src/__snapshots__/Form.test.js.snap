// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match snapshot 1`] = `
<form>
  <input
    name="title"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <input
    name="description"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    SUBMIT
  </button>
</form>
`;
